README.md
# GPTQ-triton

This is my attempt at implementing a Triton kernel for GPTQ inference.  This code is based on the [GPTQ-for-LLaMa](https://github.com/qwopqwop200/GPTQ-for-LLaMa) codebase, which is itself based on the [GPTQ](https://github.com/IST-DASLab/gptq) codebase.

```
@article{frantar-gptq,
  title={{GPTQ}: Accurate Post-training Compression for Generative Pretrained Transformers}, 
  author={Elias Frantar and Saleh Ashkboos and Torsten Hoefler and Dan Alistarh},
  year={2022},
  journal={arXiv preprint arXiv:2210.17323}
}
```

## Installation

`pip install .`


## Motivation

As of today (2023-03-27) the CUDA kernels in the aforementioned codebases do not scale well with context length, running up to 10x slower when the context is large versus the equivilent FP16 model.  To solve this I'm implementing the inference kernel in Triton, which should allow for much better scaling.

The implementation is based around the matmul tutorial from the Triton documentation.  The main difference is decoding the quantized weights before performing each sub-block of the matrix multiplication.

Fusing of the FF layers and QKV matrix are also applied.


## Performance

This benchmark was run on a 3090 using the `benchmark_generate.py` script.

![Triton benchmark graph](TritonBench.png)


## Accuracy (PPL)

The following results were obtained using the `ppl.py` script with a stride of 512 and a context length of 2048.
For the 4bit CUDA results, a custom version of `ppl.py` was used, as the current script is dedicated to the Triton kernel convensions.
it/s numbers are from a 3090.


| [LLaMA-7B](https://arxiv.org/abs/2302.13971)       | Bits | group-size | memory(MiB) | it/s | Wikitext2 |  PTB  |  C4  | 
| -------------------------------------------------- | ---- | ---------- | ----------- | ---- | --------- | ----- | ---- |
| FP16                                               |  16  |      -     |    17373    | 1.64 |    5.04   |  7.85 | 6.99 |
| GPTQ CUDA                                          |   4  |     -1     |     8805    | 0.11 |    5.44   |  8.24 |   -  |
| GPTQ Triton                                        |   4  |     -1     |     6323    | 1.70 |    5.44   |  8.24 | 7.48 |


| [LLaMA-13B](https://arxiv.org/abs/2302.13971)      | Bits | group-size | memory(MiB) | it/s | Wikitext2 |  PTB  |  C4  |
| -------------------------------------------------- | ---- | ---------- | ----------- | ---- | --------- | ----- | ---- |
| FP16                                               |  16  |      -     |    31633    |   -  |    4.52   |  7.19 | 6.66 |
| GPTQ Triton                                        |   4  |     -1     |    10325    | 0.95 |    4.74   |  7.49 | 7.00 |
| GPTQ Triton                                        |   4  |    128     |     9547    | 0.92 |    4.67   |  7.38 | 6.99 |


| [LLaMA-30B](https://arxiv.org/abs/2302.13971)      | Bits | group-size | memory(MiB) | it/s | Wikitext2 |  PTB  |  C4  |
| -------------------------------------------------- | ---- | ---------- | ----------- | ---- | --------- | ----- | ---- |
| FP16                                               |  16  |      -     |    72491    |   -  |    3.61   |  6.50 | 6.07 |
| GPTQ Triton                                        |   4  |     -1     |    19989    | 0.40 |    3.89   |  6.71 | 6.44 |
| GPTQ Triton                                        |   4  |    128     |    20055    | 0.39 |    3.79   |  6.70 | 6.34 |
| GPTQ Triton                                        |   4  |    512     |    19547    | 0.39 |    3.78   |  6.62 | 6.30 |


## Requirements

See `setup.cfg`, but note that a nightly `transformers` is preferred right now. v4.28.1 might work.  Known working `transformers` commit is `28f26c107b4a1c5c7e32ed4d9575622da0627a40`.


## Quantizing a model

The `quantize.py` script is used to quantize a HuggingFace model.  Example usage:

`./quantize.py --model <Path to a HF FP16 model> --dataset c4 --wbits 4 --groupsize -1 --act-order --true-sequential --save <Path to the output folder>`

Arguments:

* `--model`: Path to a HF FP16 model
* `--dataset`: Dataset to use for calibration.  Can be `wikitext-2`, `ptb`, `ptb-new` or `c4`.
* `--seed`: Seed for sampling the calibration data.
* `--nsamples`: Number of calibration data samples.
* `--percdamp`: Percent of the average Hessian diagonal to use for dampening (default 0.01).
* `--wbits`: Number of bits to use for quantization.
* `--groupsize`: Groupsize to use for quantization; default (-1) uses full row.
* `--save`: Save quantized result to this folder.
* `--safetensors`: Save using the safetensors format.
* `--act-order`: Use activation order quantization.
* `--true-sequential`: Use true sequential quantization.

**NOTE:** The Triton kernel is currently only implemented for 4-bits.

### Explanation of `groupsize`

The GPTQ quantization algorithm gets applied to `nn.Linear`, `nn.Conv2d`, and `transformers.Conv1d` layers.  (NOTE: `quantize.py` currently only supports LLaMA like models, and thus only `nn.Linear` layers are quantized, and `lm_head` is skipped.)  Each matrix is quantized into a quantized weight matrix, quantized zeros, and float16 scale (bias is not quantized).  During matmul, the weights are decoded using the formula `w = (w - z - 1) * s`.

Scales and zeros are per-outfeature, so when there is no grouping, scales and zeros would be `1xOutfeatures`.  That means that each row of the matrix (i.e. along the infeatures dimension) is quantized using the same scalar scale and zero.  When grouping is used, each row is split into `groupsize` values, and each group is quantized using its own scalar scale and zero.  This means that the scales and zeros are `(Infeatures//groupsize)xOutfeatures`.

Groupsize provides a tradeoff.  Lower groupsizes offer more granularity to the quantization and thus less loss of accuracy, but decrease the memory savings offered by quantization.

### Explanation of `nsamples` and `dataset`
nsamples and dataset effect the calibration data.  This input data is fed through the network during quantization to calibrate the algorithm.  See the GPTQ paper for more detail.

### Explanation of `true-sequential`
Models are quantized sequentially, one "layer" at a time.  For example, LLaMA 7B has 32 layers, starting after the input embedding, following by the head.  Each of these layers has many different Linear modules that can be quantized.  Without the `true-sequential` flag, these Linear modules will be quantized in an arbitrary order.  With the `true-sequential` flag, the Linear modules will be quantized in the order they would be encountered during a forward pass.  This can provide an accuracy boost.

### Explanation of `act-order`

I don't know.  Looking at the code (in `gptq.py`), it seems to re-order the matrix before quantization based on the `argsort` of the estimated `H`.  The order in which the matrix columns of the matrix are quantized might have an impact on final accuracy.  `act-order` was introduced by the GPTQ authors to improve accuracy when quantizing "small" models like LLaMA 7B.


## Files

* `benchmark_generate.py` - A script for benchmarking generation speed at different prompt lengths and generation lengths.

* `Benchmark.ipynb` - A notebook for benchmarking the Triton kernel against the CUDA kernel and FP16.

* `quantize.py` - A script for quantizing a model.

* `generate.py` - An example script for generating text from a model.  Example usage: `./generate.py --model <Path to your quantized model> --quant --prompt "Write a story about a duck: Once upon a time there was a duck" --temperature 0.6 --top-p 0.6 --repetition-penalty 1.1`

* `ppl.py` - A script for calculating the perplexity of a model against wikitext2, PTB, and C4.  This is useful for verifying correctness of the Triton kernel, comparing it to the CUDA kernel and the original FP16 model.

* `Verify.ipynb` - A notebook for verifying the correctness of the Triton kernel.


CHANGELOG.md
# Changelog

## [0.0.4] - 2023-04-29

- Fixed an interaction between the fused QKV projection and the key-value cache that caused excessive memory usage.


## [0.0.3] - 2023-04-28

- Disabled cache in `ppl.py`; isn't used and saves memory.
- Added more benchmarks to README.
- Fixed bug in `generate.py`; generated sequence length was not calculated correctly.


## [0.0.3] - 2023-04-19

- Added support for groupsize.
    - Note: fuse_mlp is not recommended for groupsize != -1.  It is now disabled automatically during loading if the model has grouping, unless fuse_mlp is explictly set to True.  This is a result of the current kernel implementation being slower than the naive implementation for groupsize != -1.
- Added a warning if `act_order` and `groupsize` are used together.  They are not compatible.

quantize.py
#!/usr/bin/env python3
# Based on https://github.com/IST-DASLab/gptq
# Quantize a model using the GPTQ algorithm.
import argparse
import json
from pathlib import Path
import shutil
import time
from typing import Optional

import torch
import torch.nn as nn
from datautils import get_dataset
from gptq import GPTQ, Quantizer
import gptq
from gptq_triton import QuantLinear, quant_linear
from tqdm import tqdm
from transformers import AutoTokenizer, LlamaForCausalLM


parser = argparse.ArgumentParser()

parser.add_argument('--model', type=str, required=True, help='llama model to load')
parser.add_argument('--dataset', type=str, choices=['wikitext-2', 'ptb', 'ptb-new', 'c4'], required=True, help='Where to extract calibration data from.')
parser.add_argument('--seed',type=int, default=0, help='Seed for sampling the calibration data.')
parser.add_argument('--nsamples', type=int, default=128, help='Number of calibration data samples.')
parser.add_argument('--percdamp', type=float, default=.01, help='Percent of the average Hessian diagonal to use for dampening.')
parser.add_argument('--wbits', type=int, required=True, choices=[2, 4, 8], help='#bits to use for quantization.')
parser.add_argument('--groupsize', type=int, default=-1, help='Groupsize to use for quantization; default uses full row.')
parser.add_argument('--save', type=str, required=True, help='Save quantized result to this folder.')
parser.add_argument('--safetensors', action='store_true', help='Whether to save tensors in safetensors format.')
parser.add_argument('--sym', action='store_true', help='Whether to perform symmetric quantization.')
parser.add_argument('--act-order', action='store_true', help='Whether to apply the activation order GPTQ heuristic')
parser.add_argument('--true-sequential', action='store_true', help='Whether to run in true sequential model.')


def main():
	args = parser.parse_args()
	args.save = Path(args.save)

	if args.act_order and args.groupsize != -1:
		raise ValueError('Cannot use act_order and groupsize together')

	print('Loading model...')
	model = get_llama(args.model)
	model.eval()
	tokenizer = AutoTokenizer.from_pretrained(args.model, use_fast=False)

	print('Loading data...')
	dataloader = get_dataset(args.dataset, tokenizer, nsamples=args.nsamples, seed=args.seed, seqlen=model.seqlen)

	print('Quantizing...')
	tick = time.time()
	quantizers = llama_sequential(model, dataloader, device='cuda', wbits=args.wbits, nsamples=args.nsamples, true_sequential=args.true_sequential, sym=args.sym, percdamp=args.percdamp, groupsize=args.groupsize, act_order=args.act_order)
	print(f"Total time: {time.time() - tick:.2f}s")

	print('Packing...')
	llama_pack(model, quantizers, args.wbits, args.groupsize)

	print('Saving...')
	args.save.mkdir(parents=True, exist_ok=True)

	# Save the model
	if args.safetensors:
		from safetensors.torch import save_file as safe_save
		safe_save(model.state_dict(), args.save / 'model.safetensors')
	else:
		torch.save(model.state_dict(), args.save / 'model.pt')
	
	# Write quant_config.json
	with open(args.save / 'quant_config.json', 'w') as f:
		f.write(json.dumps({
			'wbits': args.wbits,
			'groupsize': args.groupsize,
		}))

	# Copy the config
	for file in ['config.json', 'generation_config.json', 'special_tokens_map.json', 'tokenizer_config.json', 'tokenizer.model']:
		shutil.copy(args.model + '/' + file, args.save / file)
	
	print('Done.')


def get_llama(model):
	def skip(*args, **kwargs):
		pass

	# NOTE: This is a nasty hack, but it speeds up model building by a huge amount
	torch.nn.init.kaiming_uniform_ = skip
	torch.nn.init.uniform_ = skip
	torch.nn.init.normal_ = skip

	model = LlamaForCausalLM.from_pretrained(model, torch_dtype='auto')
	model.seqlen = 2048

	return model


@torch.no_grad()
def llama_sequential(model, dataloader, device, wbits: int, nsamples: int, true_sequential: bool, sym: bool, percdamp: float, groupsize: int, act_order: bool):
	# Disable caching
	use_cache = model.config.use_cache
	model.config.use_cache = False

	# Prepare
	layers = model.model.layers
	dtype = next(iter(model.parameters())).dtype
	inps = torch.zeros((nsamples, model.seqlen, model.config.hidden_size), dtype=dtype, device=device)
	outs = torch.zeros_like(inps)

	# Move the first layer to GPU
	model.model.embed_tokens = model.model.embed_tokens.to(device)
	model.model.norm = model.model.norm.to(device)
	layers[0] = layers[0].to(device)

	# Create a dummy layer that catches the input and attention mask, and then bails
	# This allows us to capture all the inputs to the first layer for the calibration data
	cache = {'i': 0, 'attention_mask': None, 'position_ids': None}
	class Catcher(nn.Module):
		def __init__(self, module):
			super().__init__()
			self.module = module
		
		def forward(self, inp, **kwargs):
			inps[cache['i']] = inp
			cache['i'] += 1
			if cache['attention_mask'] is not None:
				assert torch.all(cache['attention_mask'] == kwargs['attention_mask'])
			cache['attention_mask'] = kwargs['attention_mask']
			if cache['position_ids'] is not None:
				assert torch.all(cache['position_ids'] == kwargs['position_ids'])
			cache['position_ids'] = kwargs['position_ids']
			raise ValueError
	
	layers[0] = Catcher(layers[0])
	for batch in dataloader:
		try:
			model(batch.to(device))
		except ValueError:
			pass
	layers[0] = layers[0].module

	# Move things back to the CPU (but not the first layer, since we'll just move it back to GPU immediately below)
	model.model.embed_tokens = model.model.embed_tokens.cpu()
	model.model.norm = model.model.norm.cpu()
	torch.cuda.empty_cache()

	attention_mask = cache['attention_mask']
	position_ids = cache['position_ids']
	quantizers = {}

	# Layers are quantized in order, and only one layer lives on the GPU at a time to save memory
	# Otherwise quantizing large models would be impossible (NOTE for future readers: are you enjoying your 1TB VRAM?)
	for i, layer in tqdm(enumerate(layers), total=len(layers)):
		layer = layer.to(device)
		full = {name: m for name, m in layer.named_modules() if isinstance(m, nn.Linear)}

		if true_sequential:
			sequential = [
				['self_attn.k_proj', 'self_attn.v_proj', 'self_attn.q_proj'],
				['self_attn.o_proj'],
				['mlp.up_proj', 'mlp.gate_proj'],
				['mlp.down_proj']
			]
		else:
			sequential = [list(full.keys())]
		
		# For each subset of linear layers
		for names in sequential:
			subset = {n: full[n] for n in names}
			gptq = {}

			# Prepare a quantizer for each linear layer
			for name in subset:
				gptq[name] = GPTQ(subset[name])
				gptq[name].quantizer = Quantizer()
				gptq[name].quantizer.configure(wbits, perchannel=True, sym=sym, mse=False)
			
			# Feed data to the quantizer, and save outs
			def add_batch(name):
				def tmp(_, inp, out):
					gptq[name].add_batch(inp[0].data, out.data)
				return tmp
			
			handles = []
			for name in subset:
				handles.append(subset[name].register_forward_hook(add_batch(name)))
			for j in range(nsamples):
				outs[j] = layer(inps[j].unsqueeze(0), attention_mask=attention_mask, position_ids=position_ids)[0]  # TODO: Saving outs doesn't seem needed here?
			for h in handles:
				h.remove()

			# With the data collected, quantize the layers
			for name in subset:
				print(i, name)
				scale, zero = gptq[name].fasterquant(percdamp=percdamp, groupsize=groupsize, actorder=act_order)
				quantizers['model.layers.%d.%s' % (i, name)] = (gptq[name].quantizer, scale, zero)
				gptq[name].free()
		
		# Save outputs of the layer after quantization, so we can feed them into the next layer
		for j in range(nsamples):
			outs[j] = layer(inps[j].unsqueeze(0), attention_mask=attention_mask, position_ids=position_ids)[0]

		# Move the layer back to the CPU, and free up memory
		layers[i] = layer.cpu()
		del layer
		del gptq 
		torch.cuda.empty_cache()

		# Swap buffers
		inps, outs = outs, inps

	# Restore settings
	model.config.use_cache = use_cache

	return quantizers


def llama_pack(model, quantizers, wbits: int, groupsize: int):
	# Find all the quantized layers
	layers = {name: m for name, m in model.named_modules() if isinstance(m, nn.Linear)}
	layers = {n: layers[n] for n in quantizers}

	# Replace all applicable instances of Linear with QuantLinear in the model
	quant_linear.make_quant(model, wbits, groupsize)

	for name, m in tqdm(model.named_modules(), total=len(list(model.named_modules()))):
		if not isinstance(m, QuantLinear):
			continue

		quantizer, scale, zero = quantizers[name]
		quantizer, scale, zero = quantizer.cpu(), scale.cpu(), zero.cpu()
		pack_linear(m, layers[name].weight.data, scale, zero, m.bias)


def pack_linear(quant, weights: torch.FloatTensor, scales: torch.FloatTensor, zeros, bias: Optional[torch.FloatTensor]):
	"""
	Packs the quantized weights, scales, and zero points into a QuantLinear layer
	"""
	scales = scales.t().contiguous()
	zeros = zeros.t().contiguous()
	scale_zeros = zeros * scales

	quant.scales = scales.clone().to(torch.float16)

	if quant.bias is not None:
		quant.bias = bias.clone().to(torch.float16)
	
	# Round weights to nearest integer based on scale and zero point
	# Each weight will be one int, but should not exceed quant.bits
	intweight = []
	for idx in range(quant.infeatures):
		g_idx = idx // quant.groupsize
		# TODO: This is oddly complex.  The `gptq.quantize` function does `return scale * (q - zero)`, so shouldn't
		# this just be `q = torch.round((weights[:,idx] / scales[g_idx]) + zero[g_idx])`?
		q = torch.round((weights[:,idx] + scale_zeros[g_idx]) / scales[g_idx]).to(torch.int32)
		intweight.append(q[:,None])
	intweight = torch.cat(intweight,dim=1)
	intweight = intweight.t().contiguous()

	# Now pack the weights into uint32's
	#qweight = torch.zeros((intweight.shape[0] // 32 * quant.bits, intweight.shape[1]), dtype=torch.int32)
	quant.qweight.zero_()
	i = 0
	row = 0
	while row < quant.qweight.shape[0]:
		if quant.bits in [2,4,8]:
			for j in range(i, i + (32 // quant.bits)):
				quant.qweight[row] |= intweight[j] << (quant.bits * (j - i))
			i += 32 // quant.bits
			row += 1
		else:
			raise NotImplementedError("Only 2,4,8 bits are supported.")
	
	# Subtract 1 from the zero point
	zeros = zeros - 1

	# Pack the zero points into uint32's
	zeros = zeros.to(torch.int32)
	#qzeros = np.zeros((zeros.shape[0], zeros.shape[1] // 256 * (self.bits * 8)), dtype=np.uint32)
	quant.qzeros.zero_()
	i = 0
	col = 0
	while col < quant.qzeros.shape[1]:
		if quant.bits in [2,4,8]:
			for j in range(i, i + (32 // quant.bits)):
				quant.qzeros[:, col] |= zeros[:, j] << (quant.bits * (j - i))
			i += 32 // quant.bits
			col += 1
		else:
			raise NotImplementedError("Only 2,4,8 bits are supported.")


@torch.no_grad()
def dumbquant(layer, bits: int, groupsize: int = -1, perchannel: bool = True, sym: bool = False, mse: bool = False):
	"""
	Used to generate test data by performing a dumb quantization on the weights of a layer.
	Layer is modified in place.
	"""
	assert isinstance(layer, nn.Linear)
	quantizer = Quantizer()
	quantizer.configure(bits, perchannel=perchannel, sym=sym, mse=mse)

	W = layer.weight.data.clone()
	W = W.float()

	quantizer.find_params(W, weight=True)  # TODO: Is this needed?

	groupsize = W.shape[1] if groupsize == -1 else groupsize
	scale = []
	zero = []

	for i in range(0, W.shape[1]):
		w = W[:, i]

		if i % groupsize == 0:
			quantizer.find_params(W[:, (i):(i + groupsize)], weight=True)
			scale.append(quantizer.scale)
			zero.append(quantizer.zero)

		q = gptq.quantize(
			w.unsqueeze(1), quantizer.scale, quantizer.zero, quantizer.maxq
		).flatten()
		layer.weight.data[:, i] = q.to(layer.weight.data.dtype)

	scale = torch.cat(scale, dim=1)
	zero = torch.cat(zero, dim=1)

	return scale, zero
			

if __name__ == '__main__':
	main()